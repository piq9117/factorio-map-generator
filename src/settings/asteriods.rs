use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};

pub struct AsteriodSettings {
    pub spawning_rate: f32,
    pub max_ray_portals_expanded_per_tick: u32,
}

impl Serialize for AsteriodSettings {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("AsteriodSettings", 2)?;
        s.serialize_field("spawning_rate", &self.spawning_rate)?;
        s.serialize_field(
            "max_ray_portals_expanded_per_tick",
            &self.max_ray_portals_expanded_per_tick,
        )?;
        s.end()
    }
}
