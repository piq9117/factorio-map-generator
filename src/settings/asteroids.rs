use serde::de::{IgnoredAny, MapAccess, Visitor};
use serde::ser::SerializeStruct;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

use std::fmt;

#[derive(Debug, Clone)]
pub struct AsteroidSettings {
    pub spawning_rate: Option<f32>,
    pub max_ray_portals_expanded_per_tick: Option<u32>,
}

impl Default for AsteroidSettings {
    fn default() -> Self {
        AsteroidSettings {
            spawning_rate: Some(0.0),
            max_ray_portals_expanded_per_tick: Some(0),
        }
    }
}

impl Serialize for AsteroidSettings {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("AsteroidSettings", 2)?;
        s.serialize_field("spawning_rate", &self.spawning_rate.unwrap_or_default())?;
        s.serialize_field(
            "max_ray_portals_expanded_per_tick",
            &self.max_ray_portals_expanded_per_tick.unwrap_or_default(),
        )?;
        s.end()
    }
}

impl<'de> Deserialize<'de> for AsteroidSettings {
    fn deserialize<D>(deserializer: D) -> Result<AsteroidSettings, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct AsteroidSettingsVisitor;
        impl<'de> Visitor<'de> for AsteroidSettingsVisitor {
            type Value = AsteroidSettings;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("AsteroidSettings")
            }

            fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
            where
                V: MapAccess<'de>,
            {
                let mut spawning_rate: Option<f32> = None;
                let mut max_ray_portals_expanded_per_tick: Option<u32> = None;

                while let Some(key) = map.next_key()? {
                    match key {
                        "spawning_rate" => spawning_rate = Some(map.next_value()?),
                        "max_ray_portals_expanded_per_tick" => {
                            max_ray_portals_expanded_per_tick = Some(map.next_value()?)
                        }
                        _ => {
                            let _: IgnoredAny = map.next_value()?;
                        }
                    }
                }
                Ok(AsteroidSettings {
                    spawning_rate: spawning_rate,
                    max_ray_portals_expanded_per_tick: max_ray_portals_expanded_per_tick,
                })
            }
        }
        deserializer.deserialize_map(AsteroidSettingsVisitor)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_asteroids_settings() {
        insta::assert_json_snapshot!(AsteroidSettings::default());
    }
}
