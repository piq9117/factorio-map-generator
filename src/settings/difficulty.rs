use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};

pub struct DifficultySettings {
    pub technology_price_multiplier: Option<f32>,
    pub spoil_time_modifier: Option<f32>,
}

impl Serialize for DifficultySettings {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("DifficultySettings", 2)?;
        s.serialize_field(
            "technology_price_multiplier",
            &self.technology_price_multiplier.unwrap_or(0.001),
        )?;
        s.serialize_field(
            "spoil_time_modifier",
            &self.spoil_time_modifier.unwrap_or(0.0),
        )?;
        s.end()
    }
}
