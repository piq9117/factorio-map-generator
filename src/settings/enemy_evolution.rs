use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};
pub struct EnemyEvolutionSettings {
    pub enabled: bool,
    pub time_factor: Option<f32>,
    pub destroy_factor: Option<f32>,
    pub pollution_factor: Option<f32>,
}

impl Serialize for EnemyEvolutionSettings {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("EnemyEvolutionSettings", 4)?;
        s.serialize_field("enabled", &self.enabled)?;
        s.serialize_field("time_factor", &self.time_factor.unwrap_or(0.0))?;
        s.serialize_field("destroy_factor", &self.destroy_factor.unwrap_or(0.0))?;
        s.serialize_field("pollution_factor", &self.pollution_factor.unwrap_or(0.0))?;
        s.end()
    }
}
