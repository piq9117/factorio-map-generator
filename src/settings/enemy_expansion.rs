use serde::de;
use serde::de::{IgnoredAny, MapAccess, Visitor};
use serde::ser::SerializeStruct;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::fmt;

#[derive(Debug)]
pub struct EnemyExpansionSettings {
    pub enabled: bool,
    pub max_expansion_distance: Option<u32>,
    pub friendly_base_influence_radius: Option<u32>,
    pub enemy_building_influence_radius: Option<u32>,
    pub building_coefficient: Option<f32>,
    pub other_base_coefficient: Option<f32>,
    pub neighbouring_chunk_coefficient: Option<f32>,
    pub neighbouring_base_chunk_coefficient: Option<f32>,
    pub max_colliding_tiles_coefficient: Option<f32>,
    pub settler_group_min_size: Option<u32>,
    pub settler_group_max_size: Option<u32>,
    pub min_expansion_cooldown: Option<u32>,
    pub max_expansion_cooldown: Option<u32>,
}

impl Serialize for EnemyExpansionSettings {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("EnemyExpansionSettings", 12)?;
        s.serialize_field("enabled", &self.enabled)?;
        s.serialize_field(
            "max_expansion_distance",
            &self.max_expansion_distance.unwrap_or(0),
        )?;
        s.serialize_field(
            "friendly_base_influence_radius",
            &self.friendly_base_influence_radius.unwrap_or(0),
        )?;
        s.serialize_field(
            "enemy_building_influence_radius",
            &self.enemy_building_influence_radius.unwrap_or(0),
        )?;
        s.serialize_field(
            "building_coefficient",
            &self.building_coefficient.unwrap_or(0.0),
        )?;
        s.serialize_field(
            "other_base_coefficient",
            &self.other_base_coefficient.unwrap_or(0.0),
        )?;
        s.serialize_field(
            "neighbouring_chunk_coefficient",
            &self.neighbouring_chunk_coefficient.unwrap_or(0.0),
        )?;
        s.serialize_field(
            "neighbouring_base_chunk_coefficient",
            &self.neighbouring_base_chunk_coefficient.unwrap_or(0.0),
        )?;
        s.serialize_field(
            "max_colliding_tiles_coefficient",
            &self.max_colliding_tiles_coefficient.unwrap_or(0.0),
        )?;
        s.serialize_field(
            "settler_group_min_size",
            &self.settler_group_min_size.unwrap_or(0),
        )?;
        s.serialize_field(
            "settler_group_max_size",
            &self.settler_group_max_size.unwrap_or(0),
        )?;
        s.serialize_field(
            "min_expansion_cooldown",
            &self.min_expansion_cooldown.unwrap_or(0),
        )?;
        s.serialize_field(
            "max_expansion_cooldown",
            &self.max_expansion_cooldown.unwrap_or(0),
        )?;
        s.end()
    }
}

impl<'de> Deserialize<'de> for EnemyExpansionSettings {
    fn deserialize<D>(deserializer: D) -> Result<EnemyExpansionSettings, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct EnemyExpansionSettingsVisitor;
        impl<'de> Visitor<'de> for EnemyExpansionSettingsVisitor {
            type Value = EnemyExpansionSettings;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("EnemyExpansionSettings")
            }

            fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
            where
                V: MapAccess<'de>,
            {
                let mut enabled: Option<bool> = None;
                let mut max_expansion_distance: Option<u32> = Some(0);
                let mut friendly_base_influence_radius: Option<u32> = Some(0);
                let mut enemy_building_influence_radius: Option<u32> = Some(0);
                let mut building_coefficient: Option<f32> = Some(0.0);
                let mut other_base_coefficient: Option<f32> = Some(0.0);
                let mut neighbouring_chunk_coefficient: Option<f32> = Some(0.0);
                let mut neighbouring_base_chunk_coefficient: Option<f32> = Some(0.0);
                let mut max_colliding_tiles_coefficient: Option<f32> = Some(0.0);
                let mut settler_group_min_size: Option<u32> = Some(0);
                let mut settler_group_max_size: Option<u32> = Some(0);
                let mut min_expansion_cooldown: Option<u32> = Some(0);
                let mut max_expansion_cooldown: Option<u32> = Some(0);

                while let Some(key) = map.next_key()? {
                    match key {
                        "enabled" => enabled = Some(map.next_value()?),
                        "max_expansion_distance" => {
                            max_expansion_distance = Some(map.next_value()?)
                        }
                        "friendly_base_influence_radius" => {
                            friendly_base_influence_radius = Some(map.next_value()?)
                        }
                        "enemy_building_influence_radius" => {
                            enemy_building_influence_radius = Some(map.next_value()?)
                        }
                        "building_coefficient" => building_coefficient = Some(map.next_value()?),
                        "other_base_coefficient" => {
                            other_base_coefficient = Some(map.next_value()?)
                        }
                        "neighbouring_chunk_coefficient" => {
                            neighbouring_chunk_coefficient = Some(map.next_value()?)
                        }
                        "neighbouring_base_chunk_coefficient" => {
                            neighbouring_base_chunk_coefficient = Some(map.next_value()?)
                        }
                        "max_colliding_tiles_coefficient" => {
                            max_colliding_tiles_coefficient = Some(map.next_value()?)
                        }
                        "settler_group_min_size" => {
                            settler_group_min_size = Some(map.next_value()?)
                        }
                        "settler_group_max_size" => {
                            settler_group_max_size = Some(map.next_value()?)
                        }
                        "min_expansion_cooldown" => {
                            min_expansion_cooldown = Some(map.next_value()?)
                        }
                        "max_expansion_cooldown" => {
                            max_expansion_cooldown = Some(map.next_value()?)
                        }
                        _ => {
                            let _: IgnoredAny = map.next_value()?;
                        }
                    }
                }
                let enabled = enabled.ok_or_else(|| de::Error::missing_field("enabled"))?;

                Ok(EnemyExpansionSettings {
                    enabled: enabled,
                    max_expansion_distance: max_expansion_distance,
                    friendly_base_influence_radius: friendly_base_influence_radius,
                    enemy_building_influence_radius: enemy_building_influence_radius,
                    building_coefficient: building_coefficient,
                    other_base_coefficient: other_base_coefficient,
                    neighbouring_chunk_coefficient: neighbouring_chunk_coefficient,
                    neighbouring_base_chunk_coefficient: neighbouring_base_chunk_coefficient,
                    max_colliding_tiles_coefficient: max_colliding_tiles_coefficient,
                    settler_group_min_size: settler_group_min_size,
                    settler_group_max_size: settler_group_max_size,
                    min_expansion_cooldown: min_expansion_cooldown,
                    max_expansion_cooldown: max_expansion_cooldown,
                })
            }
        }
        deserializer.deserialize_map(EnemyExpansionSettingsVisitor)
    }
}
