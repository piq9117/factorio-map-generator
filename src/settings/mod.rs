pub mod asteroids;
pub mod difficulty;
pub mod enemy_evolution;
pub mod enemy_expansion;
pub mod map;
pub mod path_finder;
pub mod pollution;
pub mod state_steering;
pub mod steering;
pub mod unit_group;
