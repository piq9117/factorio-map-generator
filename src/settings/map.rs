use serde::de::{IgnoredAny, MapAccess, Visitor};
use serde::ser::SerializeStruct;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

use std::fmt;

use crate::settings::{
    asteroids::AsteroidSettings, difficulty::DifficultySettings,
    enemy_evolution::EnemyEvolutionSettings, enemy_expansion::EnemyExpansionSettings,
    path_finder::PathFinderSettings, pollution::PollutionSettings, steering::SteeringSettings,
    unit_group::UnitGroupSettings,
};

#[derive(Debug)]
pub struct MapSettings<'a> {
    pub difficulty_settings: Option<DifficultySettings>,
    pub pollution: Option<PollutionSettings<'a>>,
    pub steering: Option<SteeringSettings>,
    pub enemy_evolution: Option<EnemyEvolutionSettings>,
    pub enemy_expansion: Option<EnemyExpansionSettings>,
    pub unit_group: Option<UnitGroupSettings>,
    pub path_finder: Option<PathFinderSettings>,
    pub max_failed_behavior_count: Option<u32>,
    pub asteroids: Option<AsteroidSettings>,
}

impl<'a> Serialize for MapSettings<'a> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("MapSettings", 2)?;
        s.serialize_field("difficulty_settings", &self.difficulty_settings)?;
        s.serialize_field("pollution", &self.pollution)?;
        s.serialize_field("steering", &self.steering)?;
        s.serialize_field("enemy_evolution", &self.enemy_evolution)?;
        s.serialize_field("enemy_expansion", &self.enemy_expansion)?;
        s.serialize_field("unit_group", &self.unit_group)?;
        s.serialize_field("path_finder", &self.path_finder)?;
        s.serialize_field(
            "max_failed_behavior_count",
            &self.max_failed_behavior_count.unwrap_or(0),
        )?;
        s.serialize_field("asteroids", &self.asteroids)?;
        s.end()
    }
}

impl<'de> Deserialize<'de> for MapSettings<'de> {
    fn deserialize<D>(deserializer: D) -> Result<MapSettings<'de>, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct MapSettingsVisitor;
        impl<'de> Visitor<'de> for MapSettingsVisitor {
            type Value = MapSettings<'de>;
            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("MapSettings")
            }

            fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
            where
                V: MapAccess<'de>,
            {
                let mut difficulty_settings: Option<DifficultySettings> = None;
                let mut pollution: Option<PollutionSettings<'de>> = None;
                let mut steering: Option<SteeringSettings> = None;
                let mut enemy_evolution: Option<EnemyEvolutionSettings> = None;
                let mut enemy_expansion: Option<EnemyExpansionSettings> = None;
                let mut unit_group: Option<UnitGroupSettings> = None;
                let mut path_finder: Option<PathFinderSettings> = None;
                let mut max_failed_behavior_count: Option<u32> = None;
                let mut asteroids: Option<AsteroidSettings> = None;

                while let Some(key) = map.next_key()? {
                    match key {
                        "difficulty_settings" => difficulty_settings = map.next_value()?,
                        "pollution" => pollution = map.next_value()?,
                        "steering" => steering = map.next_value()?,
                        "enemy_evolution" => enemy_evolution = map.next_value()?,
                        "enemy_expansion" => enemy_expansion = map.next_value()?,
                        "unit_group" => unit_group = map.next_value()?,
                        "path_finder" => path_finder = map.next_value()?,
                        "max_failed_behavior_count" => {
                            max_failed_behavior_count = map.next_value()?
                        }
                        "asteroids" => asteroids = map.next_value()?,
                        _ => {
                            let _: IgnoredAny = map.next_value()?;
                        }
                    }
                }
                Ok(MapSettings {
                    difficulty_settings: difficulty_settings,
                    pollution: pollution,
                    steering: steering,
                    enemy_evolution: enemy_evolution,
                    enemy_expansion: enemy_expansion,
                    unit_group: unit_group,
                    path_finder: path_finder,
                    max_failed_behavior_count: max_failed_behavior_count,
                    asteroids: asteroids,
                })
            }
        }
        deserializer.deserialize_map(MapSettingsVisitor)
    }
}
