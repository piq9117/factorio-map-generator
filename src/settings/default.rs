use crate::settings::map::MapSettings;

pub fn gen_settings() -> MapSettings<'static> {
    return MapSettings {
        difficulty_settings: None,
        pollution: None,
        steering: None,
        enemy_evolution: None,
        enemy_expansion: None,
        unit_group: None,
        path_finder: None,
        max_failed_behavior_count: None,
        asteroids: None,
    };
}
