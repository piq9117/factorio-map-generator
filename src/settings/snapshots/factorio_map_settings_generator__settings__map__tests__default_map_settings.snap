---
source: src/settings/map.rs
expression: "MapSettings::default()"
---
{
  "difficulty_settings": {
    "technology_price_multiplier": 0.001,
    "spoil_time_modifier": 0.0
  },
  "pollution": {
    "enabled": false,
    "_comment_min_to_diffuse_1": "",
    "_comment_min_to_diffuse_2": "",
    "diffusion_ratio": 0.0,
    "min_to_diffuse": 0.0,
    "ageing": 0.0,
    "expected_max_per_chunk": 0.0,
    "min_to_show_per_chunk": 0.0,
    "min_pollution_to_damage_trees": 0.0,
    "pollution_with_max_forest_damage": 0.0,
    "pollution_restored_per_tree_damage": 0.0,
    "pollution_per_tree_damage": 0.0,
    "max_pollution_to_restore_trees": 0.0,
    "enemy_attack_pollution_consumption_modifier": 0.0
  },
  "steering": {
    "default": {
      "radius": 0.0,
      "separation_factor": 0.0,
      "separation_force": 0.0,
      "force_unit_fuzzy_goto_behavior": false
    },
    "moving": {
      "radius": 0.0,
      "separation_factor": 0.0,
      "separation_force": 0.0,
      "force_unit_fuzzy_goto_behavior": false
    }
  },
  "enemy_evolution": {
    "enabled": false,
    "time_factor": 0.0,
    "destroy_factor": 0.0,
    "pollution_factor": 0.0
  },
  "enemy_expansion": {
    "enabled": false,
    "max_expansion_distance": 0,
    "friendly_base_influence_radius": 0,
    "enemy_building_influence_radius": 0,
    "building_coefficient": 0.0,
    "other_base_coefficient": 0.0,
    "neighbouring_chunk_coefficient": 0.0,
    "neighbouring_base_chunk_coefficient": 0.0,
    "max_colliding_tiles_coefficient": 0.0,
    "settler_group_min_size": 0,
    "settler_group_max_size": 0,
    "min_expansion_cooldown": 0,
    "max_expansion_cooldown": 0
  },
  "unit_group": {
    "min_group_gathering_time": 0,
    "max_group_gathering_time": 0,
    "max_wait_time_for_late_members": 0,
    "max_group_radius": 0.0,
    "min_group_radius": 0.0,
    "max_member_speedup_when_behind": 0.0,
    "max_member_slowdown_when_ahead": 0.0,
    "max_group_slowdown_factor": 0.0,
    "max_group_member_fallback_factor": 0.0,
    "member_disown_distance": 0.0,
    "tick_tolerance_when_member_arrives": 0,
    "max_gathering_unit_groups": 0,
    "max_unit_group_size": 0
  },
  "path_finder": {
    "fwd2bwd_ratio": 0,
    "goal_pressure_ratio": 0.0,
    "use_path_cache": false,
    "max_steps_worked_per_tick": 0.0,
    "max_work_done_per_tick": 0,
    "short_cache_size": 0,
    "long_cache_size": 0,
    "short_cache_min_cacheable_distance": 0.0,
    "short_cache_min_algo_steps_to_cache": 0,
    "long_cache_min_cacheable_distance": 0.0,
    "cache_max_connect_to_cache_steps_multiplier": 0,
    "cache_accept_path_start_distance_ratio": 0.0,
    "cache_accept_path_end_distance_ratio": 0.0,
    "negative_cache_accept_path_start_distance_ratio": 0.0,
    "negative_cache_accept_path_end_distance_ratio": 0.0,
    "cache_path_start_distance_rating_multiplier": 0.0,
    "cache_path_end_distance_rating_multiplier": 0.0,
    "stale_enemy_with_same_destination_collision_penalty": 0.0,
    "ignore_moving_enemy_collision_distance": 0.0,
    "enemy_with_different_destination_collision_penalty": 0.0,
    "general_entity_collision_penalty": 0.0,
    "general_entity_subsequent_collision_penalty": 0.0,
    "extended_collision_penalty": 0.0,
    "max_clients_to_accept_any_new_request": 0,
    "max_clients_to_accept_short_new_request": 0,
    "direct_distance_to_consider_short_request": 0,
    "short_request_max_steps": 0,
    "short_request_ratio": 0.0,
    "min_steps_to_check_path_find_termination": 0,
    "start_to_goal_cost_multiplier_to_terminate_path_find": 0.0,
    "overload_levels": [
      0
    ],
    "overload_multipliers": [
      0.0
    ],
    "negative_path_cache_delay_interval": 0
  },
  "max_failed_behavior_count": 0,
  "asteroids": {
    "spawning_rate": 0.0,
    "max_ray_portals_expanded_per_tick": 0
  }
}
